PRT file updates

tskille opened the pull request:
Adding reservoir volume array custom FIP regions

tskille commented:
This PR has been thoroughly tested in Equinor with internal workflows that depend on using simulation results written to the PRT file from reservoir simulations. Results are good and I now believe that this PR is ready for review.

bska commented on ebos/eclwriter.hh:
```diff
@@ -243,18 +250,23 @@ public:
 
+            inplace = eclOutputModule_->calc_inplace(miscSummaryData, regionData, simulator_.gridView().comm());
+            
+            if (simulator_.gridView().comm().rank() == 0){
```
I think I'd prefer we write this check as `collectToIORank_.isIORank()` instead.

bska commented on ebos/eclwriter.hh:
```diff
@@ -243,18 +250,23 @@ public:
         std::map<std::string, std::vector<double>> regionData;
         Inplace inplace;
+        
+        if (!isSubStep)
```
I think you need to run this calculation even for substeps.  Otherwise, we'll risk breaking summary vectors like `FOIP` and `FOE`.

bska commented on ebos/eclwriter.hh:
```diff
@@ -349,14 +351,27 @@ public:
+            
+            const std::string version = moduleVersionName();
+
+            auto es_ = simulator_.vanguard().eclState();
```
This `es_` object appears to be unused.  Maybe we can just remove it?

bska commented on ebos/eclwriter.hh:
```diff
@@ -349,14 +351,27 @@ public:
+            const std::string version = moduleVersionName();
+
+            auto es_ = simulator_.vanguard().eclState();
+            auto sched_ = simulator_.vanguard().schedule();
+
+            boost::posix_time::ptime start_time = boost::posix_time::from_time_t(sched_.getStartTime());
```
Since we use the `sched_` "only" to access the start time, maybe we can write
```C++
boost::posix_time::ptime start_time = boost::posix_time::from_time_t(simulator_.vanguard(),schedule().getStartTime());
```
to avoid creating a full copy of the `Schedule` object?

bska commented on ebos/eclwriter.hh:
```diff
@@ -349,14 +351,27 @@ public:
+
+            inplace = eclOutputModule_->calc_inplace(miscSummaryData, regionData, simulator_.gridView().comm());
+
+            if (simulator_.gridView().comm().rank() == 0){
```
I think I'd prefer `collectToIORank_.isIORank()` instead of the explicit `rank() == 0` check.  They do the same thing, of course, but the `isIORank()` check is more explicit about the context.

bska commented on ebos/eclwriter.hh:
```diff
@@ -121,6 +123,9 @@ class EclWriter : public EclGenericWriter<GetPropType<TypeTag, Properties::Grid>
 
 public:
+
+    Inplace inplace_;
```
The object **probably** does not need to be `public:`, so maybe we can put it alongside the writer's other [data members](https://github.com/OPM/opm-simulators/blob/eec2fbaaa13a65f33fbf19a4c563d8be63ea50ce/ebos/eclwriter.hh#L670-L673) instead?

bska commented on opm/simulators/flow/LogOutputHelper.cpp:
```diff
@@ -306,34 +308,77 @@ fip(const Inplace& inplace,
+            
+        Scalar field_dyn_pv = 0.0;
+        
+        for (std::size_t reg = 1; reg <= inplace.max_region(name); ++reg)
```
[`Inplace::max_region()`](https://github.com/OPM/opm-common/blob/301759536b1bdf935c56891df0e7d045028699e2/src/opm/output/eclipse/Inplace.cpp#L115) is somewhat expensive, so maybe we could write this as
```C++
for (auto nreg = inplace.max_region(name), reg = 0*nreg + 1; reg <= nreg; ++reg) {
    // ...
}
```
instead, just to ensure that we don't compute `max_region()` more than once?

bska commented on opm/simulators/flow/LogOutputHelper.cpp:
```diff
@@ -306,34 +308,77 @@ fip(const Inplace& inplace,
-
-    for (std::size_t reg = 1; reg <= inplace.max_region("FIPNUM"); ++reg) {
+    
+    for (std::size_t reg = 1; reg <= inplace.max_region(); ++reg) {
```
Same here.  If we're changing this anyway, I'd prefer we compute `inplace.max_region()` only once.

bska reviewed (commented):
Thank you for taking the effort to improve report sheet compatibility.  I just noticed a few minor issues.  On a tangential note, we might conceivably defer outputting the word `AT` to the `LogOutputHelper::timeStamp()` member function.  That way there'd be less need to be very precise about the spacing in the `lbl` input argument to `EclGenericOutputBlackoilModule<>::outputTimeStamp()`.

bska commented:
jenkins build this please

tskille commented:
Perfect, thanks a lot for your comments. I believe that I have implemented all of them. I have also re-tested local workflows to make sure that these are still working. The PR is now rebased and from my end this is now ready to go in.

tskille commented:
Unfortunately, I did forget to include the tests when building flow with this PR (BUILD TESTING OFF). When including the tests, I got a compiler/linker error and I learned that I had to update one of the regression tests. 

I now have a solution for this, and I have updated the test for class LogOutputHelper.  Hence, I believe that this PR is now ready for a final review and hopefully this is now ready to go into master.

tskille commented:
jenkins build this please

bska commented:
jenkins build this please

akva2 commented:
jenkins build this please

bska reviewed (commented):
Thanks a lot for the updates.  This looks good to me now except for one potential issue with the right-alignment of the pore-volume header in the `METRIC` and/or `FIELD` unit cases.  At some point we might look into finding a way to avoid calling `moduleVersionName()` directly from `LogOutputHelper.cpp`, but for now I'm okay with adding the new build-time dependencies here.

tskille commented on opm/simulators/flow/LogOutputHelper.cpp:
```diff
@@ -882,37 +926,37 @@ void LogOutputHelper<Scalar>::
         ss << "\n                                                     ===================================\n";
         if (units.getType() == UnitSystem::UnitType::UNIT_TYPE_METRIC) {
-            ss << "                                                     :  RESERVOIR VOLUMES      M3      :\n";
+            ss << "                                                     :  RESERVOIR VOLUMES      RM3      :\n";
```
thanks, this is fixed now.

tskille commented on opm/simulators/flow/LogOutputHelper.cpp:
```diff
@@ -817,63 +862,62 @@ outputRegionFluidInPlace_(std::unordered_map<Inplace::Phase, Scalar> oip,
-        ss << name << " report region " << reg;
-    }
-
-    ss << " pressure dependent pore volume = "
-       << std::fixed << std::setprecision(0)
-       << cip[Inplace::Phase::DynamicPoreVolume] << ' '
-       << units.name(UnitSystem::measure::volume) << "\n\n";
```
I thought this could be removed since this property is also reported in reservoir volumes array 

RPTSOL
 'FIPRESV' /

```
                                                     ===================================
                                                     :  RESERVOIR VOLUMES      RM3     :
 :---------:---------------:---------------:---------------:---------------:---------------:
 : REGION  :  TOTAL PORE   :  PORE VOLUME  :  PORE VOLUME  : PORE VOLUME   :  PORE VOLUME  :
 :         :   VOLUME      :  CONTAINING   :  CONTAINING   : CONTAINING    :  CONTAINING   :
 :         :               :     OIL       :    WATER      :    GAS        :  HYDRO-CARBON :
 :---------:---------------:---------------:---------------:---------------:---------------
 :   FIELD :      262665108:       87119681:      159998829:       15546597:      102666278:
 :       1 :      131333065:       37446737:       78339731:       15546597:       52993334:
 :       2 :      131332043:       49672944:       81659099:              0:       49672944:
 :       3 :              0:              0:              0:              0:              0:
 :       4 :              0:              0:              0:              0:              0:
 ===========================================================================================
```

bska replied:
> I thought this could be removed since this property is also reported in reservoir volumes array

Ooops&ndash;yes, you're [absolutely right](https://github.com/OPM/opm-simulators/blob/3856421a8828413d524b263deeffc84444150a2a/opm/simulators/flow/LogOutputHelper.cpp#L908).  Thanks a lot for pointing that out!

bska commented:
jenkins build this please

bska reviewed (approved):
Thanks a lot for the updates and for taking the time to do this work.  I'll merge into master.
